File: LeetCode/173. Binary Search Tree Iterator/Solution.ts

 //Definition for a binary tree node.
 class TreeNode {
     val: number
     left: TreeNode | null
     right: TreeNode | null
     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
         this.val = (val===undefined ? 0 : val)
         this.left = (left===undefined ? null : left)
         this.right = (right===undefined ? null : right)
     }
 }
 

 class BSTIterator {
    private stack : TreeNode[];

    constructor(root: TreeNode | null) {
        this.stack = [];

        this.AddNodeToStack(root);
    }

    private AddNodeToStack(node: TreeNode | null | undefined){
        while(node != null){
            this.stack.push(node);
            node = node.left;
        }
    }

    private RemoveTopFromStack(): number{
        let removed : TreeNode | undefined | null = this.stack.pop();

        this.AddNodeToStack(removed?.right);

        return removed != null ? removed.val : -1;
    }

    next(): number {
        return this.RemoveTopFromStack();
    }

    hasNext(): boolean {
        return this.stack.length != 0;
    }
}

File: LeetCode/173. Binary Search Tree Iterator/Solution.cs

//Definition for a binary tree node.
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class BSTIterator
{
    private Stack<TreeNode> stack;

    public BSTIterator(TreeNode root)
    {
        stack = new Stack<TreeNode>();
        AddNodeToStack(root);
    }

    private void AddNodeToStack(TreeNode node)
    {
        while (node != null)
        {
            stack.Push(node);
            node = node.left;
        }
    }

    private TreeNode RemoveTopFromStack()
    {
        var removed = stack.Pop();

        AddNodeToStack(removed.right);

        return removed;
    }

    public int Next()
    {
        return RemoveTopFromStack().val;
    }

    public bool HasNext()
    {
        return stack.Count() != 0;
    }
}